# Import Data (Task 1-3)

In [3]:
# import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [4]:
# data set path

path = r"/Users/martin/anaconda_projects/11-02-2025 Instacart Basket Analysis"

In [5]:
# import customers data set 

df_customers = pd.read_csv(os.path.join(path, '02 data' , 'Original Data' , 'customers.csv'))

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 data' , 'Prepared Data' , 'ords_prods_merge.pkl'))

# Daten-Wrangling (Task 4)

## 4.1. Get an overview 

- To get an impression of the columns and data types
- To recognize possible problems such as missing values or mixed data types


In [8]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

##  4.2 Revise Column Names  

To improve **consistency, readability, and clarity**, the following column names have been revised:  

###  Changes Applied:
- **"First Name"** → `first_name` *(converted to snake_case)*
- **"Surnam"** → `last_name` *(corrected typo, changed to last name for clarity)*
- **"Gender"** → `gender` *(converted to lowercase for consistency)*
- **"STATE"** → `state` *(converted to lowercase for consistency)*
- **"Age"** → `age` *(converted to lowercase for consistency)*
- **"date_joined"** → *Possibly should be converted to a date format (`datetime`) for better usability.*
- **"n_dependants"** → `num_dependants` *(renamed for clarity)*
- **"fam_status"** → `family_status` *(renamed for better understanding)*
- **"income"** →  *No changes needed (already clear and consistent).*


In [10]:
df_customers.rename(columns={
    'First Name': 'first_name',
    'Surnam': 'last_name',
    'Gender': 'gender',
    'STATE': 'state',
    'Age': 'age',
    'n_dependants': 'num_dependants',
    'fam_status': 'family_status'
}, inplace=True)

In [11]:
# check whether the columns have been renamed correctly

print(df_customers.columns)

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'num_dependants', 'family_status', 'income'],
      dtype='object')


## 4.3 check & correct data types

In [13]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   user_id         206209 non-null  int64 
 1   first_name      194950 non-null  object
 2   last_name       206209 non-null  object
 3   gender          206209 non-null  object
 4   state           206209 non-null  object
 5   age             206209 non-null  int64 
 6   date_joined     206209 non-null  object
 7   num_dependants  206209 non-null  int64 
 8   family_status   206209 non-null  object
 9   income          206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


##  What Should I Adjust?  

### 1️ **Convert "date_joined" to Datetime Format**  
- Currently stored as an **object (string)**.  
- Should be converted to **datetime format** for better usability in time-based analyses.  

### 2️ **Handle Missing Values in "first_name"**  
- "first_name" has **fewer values** than other columns:  
  - **194,950** entries vs. **206,209** total rows.  
  - This indicates **missing values**.  
- If I need to use this column, I should decide how to handle the missing values:  
  - **Option 1:** Fill with an empty string (`""`).  
  - **Option 2:** Replace with `NaN` for easier filtering.  
  - **Option 3:** Remove the column if it’s not essential for analysis.  


## 4.4 Changing 'date_joined' to datetime format

In [16]:
df_customers['date_joined'] = pd.to_datetime(df_customers['date_joined'])

In [17]:
# checking if the format for 'date_joined' is now correct

df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   user_id         206209 non-null  int64         
 1   first_name      194950 non-null  object        
 2   last_name       206209 non-null  object        
 3   gender          206209 non-null  object        
 4   state           206209 non-null  object        
 5   age             206209 non-null  int64         
 6   date_joined     206209 non-null  datetime64[ns]
 7   num_dependants  206209 non-null  int64         
 8   family_status   206209 non-null  object        
 9   income          206209 non-null  int64         
dtypes: datetime64[ns](1), int64(4), object(5)
memory usage: 15.7+ MB


##  4.5 Handling the "first_name" Column  

###  Issue Identified:
- The **`first_name`** column contains **missing values**, whereas **`last_name`** remains complete.  
- Since `first_name` is **not essential for the analysis** and does not provide additional value beyond `last_name`, it can be removed.  

###  Decision & Rationale:
- **Removing unnecessary columns** helps to:  
  - **Streamline the dataset**, making it more efficient.  
  - **Reduce complexity**, ensuring cleaner and more manageable data.  
  - **Improve performance** in further analysis.  

By eliminating `first_name`, I ensure that only **relevant and complete** data is retained for processing.  


In [19]:
df_customers.drop(columns=['first_name'], inplace=True)

In [20]:
# check whether the column first_name have been removed correctly

print(df_customers.columns)

Index(['user_id', 'last_name', 'gender', 'state', 'age', 'date_joined',
       'num_dependants', 'family_status', 'income'],
      dtype='object')


###  Adjusting Column Names for Simplicity & Consistency  

As part of the process of **removing the `first_name` column**, I will **rename `last_name` to `customer_name`** to maintain:  

 **Simplicity** – A single column for customer identification.  
 **Consistency** – Aligns with other column naming conventions.  
 **Clarity** – Avoids confusion by using a broader, more descriptive term.  

This adjustment ensures a **cleaner dataset** while retaining all necessary information for further analysis.  


In [22]:
df_customers.rename(columns = {'last_name' : 'customer_name'}, inplace = True)

In [23]:
# check whether the columns have been renamed correctly

print(df_customers.columns)

Index(['user_id', 'customer_name', 'gender', 'state', 'age', 'date_joined',
       'num_dependants', 'family_status', 'income'],
      dtype='object')


# Task 5) Data quality and consistency checks

##  5.1 Check & Handle Missing Values  

###  Are There Missing Values in Other Columns?  
- We previously identified **missing values** in the `first_name` column.  
- Now, we need to check whether **other columns** also contain missing data.  

###  Next Steps:  
 Perform a **missing value check** across all columns.  
 Decide on the appropriate **handling strategy** (e.g., filling, removing, or keeping as NaN).  

Identifying and addressing missing values is essential to ensure **data integrity and reliability** in further analysis.  


In [26]:
# Check with:

df_customers.isnull().sum()

user_id           0
customer_name     0
gender            0
state             0
age               0
date_joined       0
num_dependants    0
family_status     0
income            0
dtype: int64

### Result: There are no further values missing!

###  Result: No Further Missing Values Found!  

After performing a comprehensive check, we confirmed that **no additional missing values** exist in the dataset.  

###  Key Takeaways:  
- The previously identified issue with `first_name` was addressed.  
- All remaining columns contain **complete data**, ensuring data integrity.  
- No further action is needed regarding missing values.  

With this confirmation, we can proceed confidently with the next steps in the analysis.  


In [29]:
# Check with:

df_customers.duplicated().sum()

0

###  Result: No Duplicates Found in the Dataframe!  

After performing a duplicate check, we confirmed that **there are no duplicate entries** in the dataset.  

# Task 6) Combining Customer Data with Prepared Instacart Dataset

In [32]:
# Before starting: Getting an overview of ords_prods_merge

ords_prods_merge.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           0   2539329        1    prior             1                   2   
2           0   2539329        1    prior             1                   2   
3           0   2539329        1    prior             1                   2   
4           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   ...  prices     Busiest day    Busiest days  busiest_period_of_day  \
0  ...     9.0  Regularly busy  Regularly busy         Average orders   
1  ...    12.5  Regularly busy  Regularly busy         Average orders   
2  ...     4.4  Regularly busy  Regularly busy         Average orders   
3  ...     4.7  Regularly busy  Regularly busy         Average orders   
4  ...     1.0  Regularly busy  Regularly busy         Average orders   

   max_order  loyalty_flag avg_spending spending_flag median_days_since_order  \
0         10  New Customer     6.367797   Low spender                    20.5   
1         10  New Customer     6.367797   Low spender                    20.5   
2         10  New Customer     6.367797   Low spender                    20.5   
3         10  New Customer     6.367797   Low spender                    20.5   
4         10  New Customer     6.367797   Low spender                    20.5   

    order_frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 25 columns]

###  Result: Matching `user_id` Column Found!  

Since **`ords_prods_merge`** contains a `user_id` column, just like **`df_customers`**, we can proceed with the next step.  

###  Next Steps:  
 **Step 1: Check Key Columns**  
- Ensure that `user_id` is in the correct **data type** for merging.  
- Verify that there are **no inconsistencies** in `user_id` between the datasets.  

With this confirmation, we are ready to proceed with merging the datasets for further analysis.   


##  6.1 Check Key Columns (Data Type)  

###  Why Check Data Types?  
Before merging the datasets, we need to ensure that **`user_id`** has the **same data type** in both `ords_prods_merge` and `df_customers`.  

###  Next Steps:  
 Verify the **data type** of `user_id` in both DataFrames.  
 If necessary, **convert data types** to match and avoid merge conflicts.  
 Ensure consistency to facilitate a smooth and accurate merge.  

Checking key columns is essential to prevent **errors and inconsistencies** in the data integration process.   


In [35]:
print(df_customers['user_id'].dtype)
print(ords_prods_merge['user_id'].dtype)

int64
int64


Result: Both user_id columns have the data type **int64**, so no conversion is required. I can continue directly with the merging!

In [37]:
# Shape check 'df_customers'

df_customers.shape

(206209, 9)

In [38]:
# Shape check 'ords_prods_merge'

ords_prods_merge.shape

(32434212, 25)

##  6.2 Perform Merge  

Since the `user_id` columns match in both DataFrames, we can now **combine `df_customers` with `ords_prods_merge`** to enrich our dataset with customer information.  

###  Next Steps:  
 Perform a **merge operation** using `user_id` as the key.  
 Ensure that the **merge type** (`left`, `inner`, etc.) is chosen correctly to preserve the required data.  
 Verify that the resulting dataset contains the expected number of rows and columns.  

In [40]:
df_merged = ords_prods_merge.merge(df_customers, on='user_id', how='left')

# Why how='left'?
# - This way, all orders are retained, even if not every user_id appears in the customer data.
# - If you only want to keep customer data that is also in ords_prods_merge, you could use inner - but left makes more sense here.

##  6.3 Checking After the Merge  

###  Verification Step: Were the New Columns Added?  
After merging `df_customers` with `ords_prods_merge`, we need to confirm that the new customer-related columns have been successfully integrated.  

###  Key Checks:  
 **Are the new columns from `df_customers` present** in the merged DataFrame?  
 **Is the total number of rows unchanged**, ensuring that no data was lost during the merge?  
 **Do the new columns contain expected values**, without excessive missing (`NaN`) entries?  

Verifying the merge is essential to ensure **data integrity** before proceeding with further analysis. 


In [42]:
df_merged.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           0   2539329        1    prior             1                   2   
2           0   2539329        1    prior             1                   2   
3           0   2539329        1    prior             1                   2   
4           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   ...  median_days_since_order   order_frequency_flag customer_name  gender  \
0  ...                     20.5  Non-frequent customer        Nguyen  Female   
1  ...                     20.5  Non-frequent customer        Nguyen  Female   
2  ...                     20.5  Non-frequent customer        Nguyen  Female   
3  ...                     20.5  Non-frequent customer        Nguyen  Female   
4  ...                     20.5  Non-frequent customer        Nguyen  Female   

     state  age date_joined num_dependants family_status  income  
0  Alabama   31  2019-02-17              3       married   40423  
1  Alabama   31  2019-02-17              3       married   40423  
2  Alabama   31  2019-02-17              3       married   40423  
3  Alabama   31  2019-02-17              3       married   40423  
4  Alabama   31  2019-02-17              3       married   40423  

[5 rows x 33 columns]

Result: The columns are added

###  Row Count Validation: Has the Number of Lines Remained the Same?  

After merging `df_customers` with `ords_prods_merge`, it is crucial to confirm that the **number of rows has not decreased** to ensure that no data was lost in the process.  

In [45]:
print(ords_prods_merge.shape)
print(df_merged.shape)

(32434212, 25)
(32434212, 33)


### Result: Everything is Correct!  

The merge was successful, and all key validation checks have been passed:  

### Are There Any Missing Values in the New Columns?  

In [48]:
df_merged.isnull().sum()

Unnamed: 0                       0
order_id                         0
user_id                          0
eval_set                         0
order_number                     0
orders_day_of_week               0
order_hour_of_day                0
days_since_prior_order     2078084
product_id                       0
add_to_cart_order                0
reordered                        0
_merge                           0
product_name                 28171
aisle_id                         0
department_id                    0
prices                           0
Busiest day                      0
Busiest days                     0
busiest_period_of_day            0
max_order                        0
loyalty_flag                     0
avg_spending                     0
spending_flag                    0
median_days_since_order          0
order_frequency_flag             0
customer_name                    0
gender                           0
state                            0
age                 

###  Result & Next Steps  

The merge was successful, and we have verified the presence of missing values in the dataset.  

###  Key Findings:  
1️ **No missing values** were found in the newly added columns from `df_customers` – the merge worked correctly!  
2️ **Most missing values** are in the `days_po` and `product_name` columns, which are related to orders and products.  
3️ **These NaN values are expected** and may not require immediate correction.  

###  Next Steps:  
 Decide whether to **replace NaN values** (e.g., filling `days_po` with `0` for first-time orders or replacing `product_name` with `"Unknown Product"`).  
 Alternatively, **leave NaN values as they are** if they do not affect further analysis.  

By confirming these results, we ensure **data integrity** and can proceed confidently with the next analytical steps.   


##  Handling Missing Values  

###  Decision: Keep `NaN` Values for Data Integrity  

During the data quality check, I identified **missing values** in the `days_po` and `product_name` columns. Instead of making assumptions and filling these values with arbitrary replacements, I decided to **retain them as `NaN`** for the following reasons:  

###  Why Keep `NaN` Values?  
- **Maintains Data Integrity** – Prevents introducing incorrect or misleading values.  
- **Ensures Analytical Flexibility** – `NaN` values can be easily identified and handled when needed.  
- **Avoids Potential Biases** – Filling missing values without context could distort analysis results.  

###  Future Considerations:  
- If necessary, missing values can be addressed later based on **specific analytical needs**.  
- Further exploration may determine if `days_po` missing values indicate **first-time orders** or another pattern.  
- `product_name` missing values could be replaced with `"Unknown


# Task 8) Export Data

In [52]:
# Export data to pkl
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_combined.pkl'))